Landing support for the LLVM AOT backend on Windows

benvanik opened the pull request:
The primary change here is around how linkers are defined in the compiler and how we use them. There is now a LinkerTool type that allows for target-specific manipulation of the module prior to codegen required for making valid Windows DLLs and can be useful for things like wasm where we may need to expose additional control symbols.

The existing linking behavior is moved into UnixLinkerTool (using ld/ld.lld) while the new Windows support is in (creatively) WindowsLinkerTool. TODOs are in place for mac mach-o and wasm linking, and the UnixLinkerTool could use a lot of work (left for someone who knows more about ld).

I included instructions for the Windows builds and tried to make it as default as possible ("vcvarsall to setup your env then build with the right triple"). Still likely some tuning that could be done to reduce configuration issues but I feel like the UnixLinkerTool needs much more love first :)

Fixes #2645 by providing defaults and useful error messages when not set right.

benvanik committed:
Reworking LLVM AOT linking to pick a linker tool based on target env.
This adds the WindowsLinkerTool that uses lld-link or link to generate
the DLLs/PDBs. The existing behavior is modeled in UnixLinkerTool that
is meant for use with ld/ld.lld.

TODOs are in place for mac mach-o and wasm linking, and the UnixLinkerTool
could use a lot of work (left for someone who knows more about ld).

benvanik committed:
Adding docs for Windows + AOT.

ScottTodd committed:
Small fixes for LLVM AOT backend on Windows. (#3786)

* Fix handling when IREE_LLVMAOT_LINKER_PATH is not set.

* Intentionally leak dylib libraries on exit when tracing.

benvanik committed:
Fixing mismatched class/struct on fwd declarations.

ScottTodd reviewed (approved):
Code LGTM, and mostly ready on the branch merging side of things. Let's figure out the failing tests though.

benvanik committed:
winky boi

benvanik committed:
Disabling integrations/tensorflow/e2e LLVM JIT tests.
The LLVM JIT driver is deprecated and going away. These tests should
be using LLVM AOT instead (#2673), but this test infra is scary.
Ideally, this infra would barely be doing any of this per-backend;
there should be few - if any - things exclusively tested by this
path that aren't also covered in core tests that don't build/run
from python. Or at least, the infra should not be doing all of this
via build files (bazel is not a configuration language).

Unblocks #3757 and #3843.
